### PR TITLE
pulse backend string model parser

### LINKED ISSUE #53: Model string representation parsing functionality from qiskit aer
### What is the expected behavior?

Taking/modifying code from Aer, write a function with signature:
- Inputs:
    - Hamiltonian string representation
    - Subsystem list (list of subsystems in the model to keep when parsing)
- Outputs:
    - Hamiltonian `static_operator` and `operators` (i.e. operator data required to instantiate a `HamiltonianModel` or `Solver`)
    - List of channel names (as strings, e.g. `['D0', 'D2', 'U0']` corresponding to the operators in `operators`.

### PR DESCRIPTION
**Pull request by DanPuzzuoli**
### Summary

Closes #53.

Bring string model parser from Aer into dynamics. Currently a work in progress as I attempt to reduce the original number of files and lines of code involved.

### CODE REVIEW

**Comment by DanPuzzuoli** [comment]
This PR is ready for review. I think it adds the "smallest usable unit" that makes sense for parsing the pulse backend model strings.

All non-test files added are in `qiskit_dynamics/pulse/string_model_parser`. The main function, and probably the only one that should be used, is `string_model_parser.parse_hamiltonian_dict`, which takes in the backend hamiltonian dictionary, and returns:
- The static hamiltonian,
- A hamiltonian operators (i.e. the terms with time-dependent coefficients in the form of channels/signals), and
- A corresponding list of channel labels for the terms with time-dependent coefficients.

In subsequent developments, the operators can be used to construct a model or solver in dynamics, and the list of channels will be used in the schedule -> signal conversion routine put the output signals in the right order.

Some notes:
- The `legacy_parser.py` file compartmentalizes the actual string parsing part from the Aer pulse simulator, which is untested and that is treated as a black box. At this stage I think there are diminishing returns to trying to understand/test it, and putting it in this file is an attempt to make it clear that we are purposefully compartmentalizing it.
- I've tried to document the expected input format in the doc string for `parse_hamiltonian_dict`. I think this doc string is important to review.
- I've put everything into it's own subfolder since it consists of several files but there's really only one meaningful function coming out of it. I'd be interested in comment on this structure as well.

**Review by brosand — COMMENTED** [review]
LGTM, happy to have a longer discussion to be more detailed, but nothing jumps out.

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+                # if channel string doesn't contain anything other than channel character
+                if len(channel_str) == 1:
+                    raise QiskitError(malformed_text)
+
+                # if starts with opening bracket, verify that it ends with closing bracket
+                if channel_str[1] == "{":
+                    if not channel_str[-1] == "}":
+                        raise QiskitError(malformed_text)
+                # otherwise verify that the remainder of terms are only contains digits
```
Comment (brosand) [inline qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:292]: verify that the remainder of the terms only contain digits

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+                raise QiskitError(malformed_text)
+
+            # if first entry in channel string isn't a valid channel character
+            if channel_str[0] not in CHANNEL_CHARS:
+                raise QiskitError(malformed_text)
+
+            # Verify either that: all remaining characters are digits, or,
+            # if term starts with _SUM[ and ends with ], all remaining characters
+            # are either digits, or starts and ends with {}
+            if term[-1] == "]" and len(term) > 5 and term[:5] == "_SUM[":
```
Comment (brosand) [inline qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:280]: Shouldn't we do checks like these even on strings without channels? (specifically line 280)

File: test/dynamics/pulse/test_string_model_parser.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+
+from qiskit_dynamics.type_utils import to_array
+
+from ..common import QiskitDynamicsTestCase
+
+
+class TestHamiltonianPreParseExceptions(QiskitDynamicsTestCase):
+    """Tests for preparse formatting exceptions."""
+
+    def test_no_h_str(self):
```
Comment (brosand) [inline test/dynamics/pulse/test_string_model_parser.py:36]: Super unimportant but just to cover edge case of `h_str= ""`, add a line for that as well, since we have the code for it above. (Let me know if this is too unimportant, still trying to figure out best practices here)

File: test/dynamics/pulse/test_string_model_parser.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+        self.assertTrue("requires a non-empty 'h_str'" in str(qe.exception))
+
+    def test_empty_qub(self):
+        """Test qub dict empty raises error."""
+
+        with self.assertRaises(QiskitError) as qe:
+            hamiltonian_pre_parse_exceptions({"h_str": ["a * X0|||D0"]})
+        self.assertTrue("requires non-empty 'qub'" in str(qe.exception))
+
+    def test_too_many_vertical_bars(self):
```
Comment (brosand) [inline test/dynamics/pulse/test_string_model_parser.py:50]: Add a test for too few vertical bars (1)

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+    """Raises exceptions for improperly formatted or unsupported elements of
+    hamiltonian dict specification.
+
+    Parameters:
+        hamiltonian_dict: Dictionary specification of hamiltonian.
+    Returns:
+    Raises:
+        QiskitError: If some part of the Hamiltonian dictionary is unsupported or invalid.
+    """
+
```
Comment (brosand) [inline qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:241]: Some other ideas for preparsing if we want to (maybe unnecessary)

- check if variables exist which aren't defined (probably handled by exceptions from the evaluation).
- Check for non-alphanumeric characters that don't have predefined uses (unless those are fine as variable names)

**Review by brosand — COMMENTED** [review]
Looking good, mostly just left a couple small suggestions for testing and parsing, and a couple small typos.

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+    Returns:
+        np.ndarray corresponding to the specified operator.
+    """
+
+    # construct single system operator
+    out = single_operator_from_string(op_label, subsystem_dims[subsystem_index])
+
+    # sort subsystem labels and dimensions
+    sorted_subsystem_keys, sorted_subsystem_dims = zip(
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:67]: Basically this line is set up to sort two lists simultaneously, according to the ordering of the first list. Working inwards out:
1. The first zip combines the list into a single iterator whose elements are tuples with one element from each list (in the original order).
2. `sorted` then sorts this new iterator according to the first element in the tuples (the first list).
3. I'm not fully sure of exactly what this syntax implies, but `zip(*x)` for `x = [(a1, b1), (a2, b2), ... , (ak, bk)]` returns `[a1, ..., ak], [b1, ..., bk]`. (i.e. now that things are sorted, we split the zipped thing back into individual lists.)

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        dim: Dimension of operator.
+
+    Returns:
+        np.ndarray
+
+    Raises:
+        QiskitError: If op_label doesn't correspond to a known operator.
+    """
+
+    op_func = __operdict.get(op_label, None)
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:140]: Moved it above.

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    "I": ident,
+}
+
+
+def dag(op: np.ndarray) -> np.ndarray:
+    """Apply dagger."""
+    return np.conjugate(np.transpose(op))
+
+
+def apply_func(name: str, op: np.ndarray) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:168]: This is what existed in the original pulse simulator string parser, so just brought that over. The parser still requires this format so I will stick with this and we can add things over time if necessary.

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    * ``'qub'``: Dictionary giving subsystem dimensions. Keys are subsystem labels,
+      values are their dimensions.
+    * ``'vars'``: Dictionary with variables appearing in the terms in the ``h_str`` list,
+      keys are strings giving the variables, values are the values of the variables.
+
+    The optional argument ``subsystem_list`` specifies a subset of subsystems to keep when parsing.
+    If ``None``, all subsystems are kept. If ``subsystem_list`` is specified, then terms
+    including subsystems not in the list will be ignored.
+
+    Entires in the list ``hamiltonian_dict['h_str']`` must be formated as a product of
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:53]: Done.

[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:53]: Done.

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py (hunk 1 above)
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:293]: In this case they're not digits, they'll be a string of characters and refer to the summation index. I think adding a check for this may be more complicated than its worth.

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py (hunk 2 above)
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:281]: It appears all of these checks are based around validating channel specification - nothing else is being checked here. I think I just did this here because the legacy parser has some of its own checking for the validity of the operator part of the string, but has less checking for the validity of the channel part.

File: test/dynamics/pulse/test_string_model_parser.py (hunk 3 above)
Comment (DanPuzzuoli) [reply test/dynamics/pulse/test_string_model_parser.py:36]: Added two additional checks to this test to verify that the same error gets raised if the argument is: `{'h_str': []}` or `{'h_str': ['']}`.

This is good to point - obviously we can't be overly exhaustive of all possible validation issues but it's good to make sure nothing breaks here.

File: test/dynamics/pulse/test_string_model_parser.py (hunk 4 above)
Comment (DanPuzzuoli) [reply test/dynamics/pulse/test_string_model_parser.py:50]: Added below this one.

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py (hunk 5 above)
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:242]: These are good ideas, though interestingly these problems are actually automatically handled gracefully by virtue of how these expressions are evaluated. Since `exec` is called on the coefficient strings, if a variable isn't defined a `NameError` is raised for that variable name. I think the same applies to the second point, and the legacy parser will handle undefined operator names (via calls to `operator_from_string`).

**Comment by DanPuzzuoli** [comment]
Not sure why I can't reply to the one comment about adding the output to one of the examples in the docs. 

I'm kind of on the fence about this - two of the outputs of the function consist of arrays and so showing them isn't necessarily too informative. Given that this function is primarily for internal use anyway I'm leaning towards not adding this.

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py (hunk 1 above)
Comment (brosand) [reply qiskit_dynamics/pulse/string_model_parser/string_model_parser.py]: yes sorry by digits I meant non-weird chars, but agreed will resolve

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py (hunk 5 above)
Comment (brosand) [reply qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:242]: great this was my hope

**Comment by brosand** [comment]
> Not sure why I can't reply to the one comment about adding the output to one of the examples in the docs.
> 
> I'm kind of on the fence about this - two of the outputs of the function consist of arrays and so showing them isn't necessarily too informative. Given that this function is primarily for internal use anyway I'm leaning towards not adding this.

Not sure why I can't reply to this one either haha, but this makes sense,

File: test/dynamics/pulse/test_string_model_parser.py [hunk 10]
```diff
@@ -54,6 +62,13 @@ def test_too_many_vertical_bars(self):
             hamiltonian_pre_parse_exceptions({"h_str": ["a * X0|||D0"], "qub": {0: 2}})
         self.assertTrue("does not conform" in str(qe.exception))
 
+    def test_single_vertical_bars(self):
+        """Test that too many vertical bars raises error."""
```
Comment (brosand) [inline test/dynamics/pulse/test_string_model_parser.py:66]: should be "test that too few vertical bars"

**Review by brosand — DISMISSED** [review]
LGTM, thanks for doing those small changes

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+def dag(op: np.ndarray) -> np.ndarray:
+    """Apply dagger."""
+    return np.conjugate(np.transpose(op))
+
+
+# pylint: disable=invalid-name
+__funcdict = {"dag": dag}
+
+
+def apply_func(name: str, op: np.ndarray) -> np.ndarray:
```
Comment (chriseclectic) [inline qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:172]: This function seems sus. You could just replace it in the parser code (which is the only spot it looks to be used) with something like 
```python
elif token.type in ["Func", "Ext"]:
     if token.name == "dag":
         stack.append(np.conjugate(np.transpose(stack.pop(-1)))
     else:
          stack.append(stack.pop(-1))
```
(possibly something even simpler if you know whats going on in that code block)

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+        subsystem_list = subsystem_list.copy()
+        subsystem_list.sort()
+
+    # force keys in hamiltonian['qub'] to be ints
+    qub_dict = {int(key): val for key, val in hamiltonian_dict["qub"].items()}
+
+    subsystem_dims = {int(qubit): qub_dict[int(qubit)] for qubit in subsystem_list}
+
+    # Parse the Hamiltonian
+    system = legacy_parser(
```
Comment (chriseclectic) [inline qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:168]: No need for this function, it could just be `system = HarmiltonianParser(args).parse(other_args)`

File: qiskit_dynamics/pulse/string_model_parser/legacy_parser.py [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+    Var=re.compile(r"[a-z]+[0-9]*"),
+    Num=re.compile(r"[0-9.]+"),
+    MathOrd0=re.compile(r"[*/]"),
+    MathOrd1=re.compile(r"[+-]"),
+    BrkL=re.compile(r"\("),
+    BrkR=re.compile(r"\)"),
+)
+
+
+class HamiltonianParser:
```
[the remark this answers is not in the document]
Comment (chriseclectic) [reply qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:50]: Might also want to call this `_HamiltonianParser` just to add further indication this shouldn't be used and is internal only

**Review by chriseclectic — CHANGES_REQUESTED** [review]
Since majority of this is intended as internal code hidden from users i would recommend adding some comments to that effect in the module docstrings, and renaming most of the classes and functions to start with an underscore to further indicate they are internal methods that should not be used directly

File: qiskit_dynamics/pulse/__init__.py [hunk 14]
```diff
@@ -75,6 +75,8 @@
    :toctree: ../stubs/
 
    InstructionToSignals
+   parse_hamiltonian_dict
 """
 
 from .pulse_to_signals import InstructionToSignals
+from .string_model_parser.string_model_parser import parse_hamiltonian_dict
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/__init__.py:82]: You're right, I'll remove this.

File: qiskit_dynamics/pulse/string_model_parser/legacy_parser.py [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+
+"""Legacy code for parsing operator strings."""
+
+import re
+import copy
+from collections import namedtuple, OrderedDict
+from .operator_from_string import operator_from_string, apply_func
+
+
+def legacy_parser(operator_str, subsystem_dims, subsystem_list):
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:24]: I'm using "legacy" here to imply that it's a non-trivial piece of code that is being preserved but not "maintained" per se. I like the usage of "legacy" to imply this - I think the way you're interpreting it is basically synonymous with "expected to be deprecated". 

That all being said I'll change this if it's a sticking point for you.

File: qiskit_dynamics/pulse/string_model_parser/legacy_parser.py [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+
+    system = HamiltonianParser(h_str=operator_str, subsystem_dims=subsystem_dims)
+    system.parse(subsystem_list)
+
+    return system.compiled
+
+
+Token = namedtuple("Token", ("type", "name"))
+
+str_elements = OrderedDict(
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:35]: Moved this into the class.

File: qiskit_dynamics/pulse/string_model_parser/legacy_parser.py [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+    def __init__(self, h_str, subsystem_dims):
+        """Create new quantum operator generator
+
+        Parameters:
+            h_str (list): list of Hamiltonian string
+            subsystem_dims (dict): dimension of subsystems
+        """
+        self.h_str = h_str
+        self.subsystem_dims = {int(label): int(dim) for label, dim in subsystem_dims.items()}
+        self.__td_hams = []
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:62]: No reason, it was in the original code. I've changed it to just `td_hams`.

File: qiskit_dynamics/pulse/string_model_parser/legacy_parser.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:50]: No reason - this is part of the "legacy" aspect of it and why I wrote the `legacy_parser` function. Turning it directly into a function was non-trivial given the handling of internal state. The `legacy_parser` function was added to wrap this in the preferred standard function interface.

File: qiskit_dynamics/pulse/string_model_parser/legacy_parser.py [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+import copy
+from collections import namedtuple, OrderedDict
+from .operator_from_string import operator_from_string, apply_func
+
+
+def legacy_parser(operator_str, subsystem_dims, subsystem_list):
+    """Function wrapper for legacy parsing object."""
+
+    system = HamiltonianParser(h_str=operator_str, subsystem_dims=subsystem_dims)
+    system.parse(subsystem_list)
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:28]: Changed it to return as suggested and simplified `legacy_parser` function to one line.

File: qiskit_dynamics/pulse/string_model_parser/legacy_parser.py [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+        self.__td_hams = []
+        self.__tc_hams = []
+        self.__str2qopr = {}
+
+    @property
+    def compiled(self):
+        """Return Hamiltonian terms."""
+        return self.__tc_hams + self.__td_hams
+
+    def parse(self, qubit_list=None):
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:71]: You're right - just removed the `compiled` property and made them local variables.

File: qiskit_dynamics/pulse/string_model_parser/legacy_parser.py [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+            subsystem_dims (dict): dimension of subsystems
+        """
+        self.h_str = h_str
+        self.subsystem_dims = {int(label): int(dim) for label, dim in subsystem_dims.items()}
+        self.__td_hams = []
+        self.__tc_hams = []
+        self.__str2qopr = {}
+
+    @property
+    def compiled(self):
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:67]: Yep - removed.

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+    "Sm": a,
+    "Sp": adag,
+    "C": adag,
+    "N": N,
+    "O": N,
+    "I": ident,
+}
+
+
+def single_operator_from_string(op_label: str, dim: int) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:139]: yep - removed.

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+# pylint: disable=invalid-name
+
+"""Generate operators from string."""
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:16]: Added a warning to this and every other file in this submodule.

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+        - `'C'`, or `'Sp'`: If two dimensional, sigma plus operator, and if greater,
+          generalizes to the creation operator.
+        - `'N'`, or `'O'`: The number operator.
+        - `'I'`: The identity operator.
+
+    Note that the ordering of tensor factors is reversed.
+
+    Args:
+        op_label: The string labelling the single system operator.
+        subsystem_index: Index of the subsystem to apply the operator.
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:56]: This subsystem index is just a single `int` indicating the label of the subsystem that the operator being constructed acts on. The `subsystem_dims` argument is a dictionary mapping subsystem labels to dimensions, so that this function knows where and what order to tensor on identities. The subsystem labels in `subsystem_dims` are assumed to be ints by this function.

To address this:
- Made the type hint for `subsystem_labels` more detailed to show that both the keys and labels are ints.
- Changing `subsystem_index` to `subsystem_label` to avoid confusion that "index" might be indexing a list of subsystems (as opposed to specifying a specific label).

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+    return ident(dim) - 2 * N(dim)
+
+
+def ident(dim: int) -> np.ndarray:
+    """Identity operator."""
+    return np.eye(dim, dtype=complex)
+
+
+# operator names
+__operdict = {
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:124]: So that it isn't constructed every time the function is called.

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py (hunk 11 above)
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/operator_from_string.py:172]: yeah you're right - I think maybe it was like this in the original pulse simulator to enable further expansion of functionality later, but for now it's pointless. I've implemented your suggestion and removed `apply_func` and related things, though changed the else statement to raise an error.

File: qiskit_dynamics/pulse/string_model_parser/string_model_parser.py (hunk 12 above)
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/string_model_parser.py:168]: Unless you think otherwise I'd rather keep the `legacy_parser` function even though this is all it does as its the preferred interface (a function call).

File: qiskit_dynamics/pulse/string_model_parser/legacy_parser.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/legacy_parser.py:24]: Updated the function to have type hints.

**Comment by DanPuzzuoli** [comment]
@chriseclectic I think I've addressed most comments with the latest changes. The remaining things to discuss are:
- Usage of the word "legacy". Maybe we can change the file to `regex_parser.py` and this function to `regex_parser` - this will drop the problematic terminology and may be a more informative name (indicating this is where the true "parsing" happens).
- Using underscores for functions/classes that are meant to be internal: I'd prefer not to do this as I feel like it is unnecessary and I don't really like the way it looks. I think the fact that these things aren't being imported into the standard packaging structure/API, along with the warnings in the files, should be sufficient to imply that they are internal?

File: qiskit_dynamics/pulse/string_model_parser/operator_from_string.py [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+    # get subsystem location in ordered list
+    subsystem_location = sorted_subsystem_keys.index(subsystem_index)
+
+    # tensor identity on right if subsystem_index is not first subsystem
+    if subsystem_location != 0:
+        total_dim = np.prod(sorted_subsystem_dims[:subsystem_location])
+        out = np.kron(out, ident(total_dim))
+
+    # tensor identity on left if subsystem_index not the last subsystem
+    if subsystem_location + 1 != len(sorted_subsystem_keys):
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit_dynamics/pulse/string_model_parser/operator_from_string.py]: Changed it to utilize this.

**Comment by DanPuzzuoli** [comment]
Closing this PR as it is extremely out of date - will reopen a new one.
